### DISCUSSION TITLE
Difference between MacOS and MacCatalyst

### DISCUSSION THREAD

**Discussion by TopperDEL** [unoplatform/uno#14709] [Q&A]
I'm not very deep in the apple ecosystem, so I'm asking myself: what is the difference between MacOs and MacCatalyst?
Especially regarding Uno: should I focus on Catalyst? What "platform" is supported most?

I had some trouble trying one of my apps on a Mac where it seemed to be running better in MacOs then on Catalyst. Is this still the case? Will it get better since the sources of WinUI have been released by Microsoft?

Would be happy to get a short overview - especially regarding the feature set and which one to focus. Thanks!

**Comment by jeromelaban** [comment]
At this time, the support for macCatalyst is a lot better (production ready) when compared to AppKit support. At this time, as MAUI does not support `net8.0-macos`, most of the issues are not taken care of by Microsoft.

With regards to differentes between AppKit and Catalyst, Catalyst is essentially UIKit running on mac, with some smaller variants and a stronger sandbox. As far as Uno is concerned, we're making specific changes to Catalyst support as needed, but it's considered as stable as iOS.

**Reply by TopperDEL** [reply]
Thanks! So I should not expect to have issues around Pointer-Events anymore?

**Reply by jeromelaban** [reply]
There may be differences, but it you find any, let us know.

**Reply by TopperDEL** [reply]
Ok, I'm not at a point where I can start and debug my Mac Catalyst-Head. Yay!
So I can say: PointerMoved is not working - at least on a SKXamlCanvas. It is only working if the pointer is pressed during move.

**Reply by TopperDEL** [reply]
Other issues I have with Mac Catalyst:
- TextBoxes and some other layout controls (have to investigate more) have different default widths then on Windows. They stretch to the whole screen while they are much smaller on Windows.
- OnKeyDown does not seem to work on an inherited TextBox-control.
- OnSurfacePointerWheelChanged does not work
- WebView2 is crashing
- ProgressRing is not rendering

A lot to do. :) If you have existing issues in your head, that point to one of these issues, let me know. I'll otherwise try to pin them down with samples over the next days and create issues for them. These are unfortunately show-stoppers for my app.

**Reply by jeromelaban** [reply]
Thanks for the update!

> TextBoxes and some other layout controls (have to investigate more) have different default widths then on Windows. They stretch to the whole screen while they are much smaller on Windows.

This may be. There could be some differences with the default WinUI templates that we have not noticed.

> OnKeyDown does not seem to work on an inherited TextBox-control.

This is likely a system behavior that we may not be able handle as the native texbox may intercept it first. Still, you can open an issue for this, with the scenario you're aiming for.

> OnSurfacePointerWheelChanged does not work

This is likely related to the support of mouse pointers, in relation to https://github.com/unoplatform/uno/issues/12467.

> WebView2 is crashing

This is curious, https://github.com/unoplatform/uno/issues/9877 is known, but not crashing.

> ProgressRing is not rendering

Do you have Uno.WinUI.Lottie in your app packages? This could explain the behavior. Otherwise, take a look at the app's logs.

**Reply by TopperDEL** [reply]
Thanks @jeromelaban - once again - for the quick and detailed response!

Regarding the templates: can I compare them anywhere by myself and maybe create a PR? Where would I have to start?

Regarding the TextBox: I draw my own TextBox using another [OpenSource library](https://github.com/toptensoftware/RichTextKit). And for that I need to be in the tab order (to get the focus) and to receive keypresses on my control. So I derived from TextBox and just replace the control-template to render my own stuff. I just tried to inherit from a UserControl instead, that did not work. But it was just a quick test. I did not get OnKeyDown there on Windows, too. So I may have to investigate more how to hook into those events.

#12467 seems to be an important one for me, still. :)

Regarding WebView2: It is crashing as I needCoreWebView2.Environment to send a POST-Message to an URL. This is not implemented in Uno. [I've created an issue for that](https://github.com/unoplatform/uno/issues/14863).

The ProgressRing now renders after adding the Lottie-Package - thanks. But now it has an extra little quadratic background behin it, that I'm unable to make transparent. The [documentation](https://platform.uno/docs/articles/features/progressring.html?q=progress%20ring#using-the-microsoftuixamlcontrolsprogressring) seems a bit confusing, as it mentiones two different ProgressRings (WUX and MUX) but both reside in the exact same namespace.

**Reply by jeromelaban** [reply]
> Regarding the templates: can I compare them anywhere by myself and maybe create a PR? Where would I have to start?

Thanks for suggesting a contribution. You can find the [templates here](https://github.com/unoplatform/uno.templates/tree/main/src/Uno.Templates/content/unoapp). We may be able adjust what's missing for this specific case.

> https://github.com/unoplatform/uno/issues/12467 seems to be an important one for me, still. :)

Agreed, this is important.

> The ProgressRing now renders after adding the Lottie-Package - thanks. But now it has an extra little quadratic background behin it, that I'm unable to make transparent.

This could be a SkiaSharp issue that was not fixed specifically for Catalyst. There were similar issues for iOS/Android, but in general Catalyst issues are fixed at the same time as iOS.

> The [documentation](https://platform.uno/docs/articles/features/progressring.html?q=progress%20ring#using-the-microsoftuixamlcontrolsprogressring) seems a bit confusing, as it mentiones two different ProgressRings (WUX and MUX) but both reside in the exact same namespace.

Indeed, the documentation is misleading as the namespaces are the same, ~~we'll adjust that~~ (this will be fixed during the next release, now that the whole uno tree is using WinUI first).
